### ISSUE TITLE
RuntimeError: expected scalar type BFloat16 but found Float

### ISSUE THREAD

**Issue by brycedrennan** [brycedrennan/imaginAIry#13]
> HI Just thought I'd let you know I'm getting the same thing on PoP OS:
> 
> File "/home/jasper/anaconda3/lib/python3.9/site-packages/torch/nn/functional.py", line 2516, in group_norm return torch.group_norm(input, num_groups, weight, bias, eps, torch.backends.cudnn.enabled)RuntimeError: expected scalar type BFloat16 but found Float
> 

https://www.reddit.com/r/StableDiffusion/comments/xdcgh0/comment/ip5m1hh/?context=3

**Comment by brycedrennan** [comment]
Possibly related:
https://huggingface.co/CompVis/stable-diffusion-v1-4/discussions/42
https://www.gabriel.urdhr.fr/2022/08/28/trying-to-run-stable-diffusion-on-amd-ryzen-5-5600g/

**Comment by wseagar** [comment]
Hey @brycedrennan just for background on my older nvidia GPU I get this error too so I always run with `precision !== "autocast"` and it fixes it. I'm not sure the reasoning why.

**Comment by brycedrennan** [comment]
@wseagar a fix I have low confidence in has been deployed with imaginairy 1.6.0 
Let me know if it works.

**Comment by iscopan** [comment]
That first error was solved, but in the end this happened

`Generating 🖼  : "X" 512x512px seed:787666775 prompt-strength:7.5 steps:40 sampler-type:pl
ms
    PLMS Sampler: 100%|█████████████████████████████████████████████████████████████████| 40/40 [37:05<00:00, 55.64s/it]
/usr/lib/python3/dist-packages/apport/report.py:13: DeprecationWarning: the imp module is deprecated in favour of importlib and slated for removal in Python 3.12; see the module's documentation for alternative uses
  import fnmatch, glob, traceback, errno, sys, atexit, locale, imp, stat
Traceback (most recent call last):
  File "/home/stdiff/.local/bin/imagine", line 8, in <module>
    sys.exit(imagine_cmd())
  File "/usr/lib/python3/dist-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/stdiff/.local/lib/python3.10/site-packages/imaginairy/cmds.py", line 223, in imagine_cmd
    imagine_image_files(
  File "/home/stdiff/.local/lib/python3.10/site-packages/imaginairy/api.py", line 114, in imagine_image_files
    for result in imagine(
  File "/home/stdiff/.local/lib/python3.10/site-packages/imaginairy/api.py", line 292, in imagine
    x_sample.cpu().numpy(), "c h w -> h w c"
TypeError: Got unsupported ScalarType BFloat16`

**Comment by Littlecowherd** [comment]
some related discussions here
https://github.com/CompVis/stable-diffusion/pull/47

**Comment by brycedrennan** [comment]
That's great news @iscopan ! that new error looks pretty straightforward to resolve

**Comment by brycedrennan** [comment]
@Littlecowherd thanks I read that discussion and it looks like I'm taking a different approach to the problem.  

These issues arise from  trying to do math on numbers of different types.  The general reason we have numbers of different types is that for some situations it can improve performance to use less accurate float16 numbers.   It seems other people have resolved this by just using full float32 numbers everywhere, which could be bad for performance.

Pytorch is supposed to intelligently handle the math between float16/float32 but there was a bug in this case.  I implemented a monkeypatch to fix that bug: https://github.com/pytorch/pytorch/pull/81852

**Comment by brycedrennan** [comment]
@iscopan I've deployed 1.6.1 which should address the new problem you've identified.  🤞 that this gets us all the way

https://pypi.org/project/imaginAIry/

**Comment by iscopan** [comment]
Previous error is fixed, but it has happened again later

```
Generating 🖼  : "X" 512x512px seed:123411790 prompt-strength:7.5 steps:40 sampler-type:pl
ms
    PLMS Sampler: 100%|██████████| 40/40 [XX:XX<00:00, XX.XXs/it]
Downloading: 100%|██████████| 342/342 [00:00<00:00, 351kB/s]
Downloading: 100%|██████████| 4.44k/4.44k [00:00<00:00, 4.13MB/s]
Downloading: 100%|██████████| 1.13G/1.13G [00:14<00:00, 81.5MB/s]

/usr/lib/python3/dist-packages/apport/report.py:13: DeprecationWarning: the imp module is deprecated in favour of importlib and slated for removal in Python 3.12; see the module's documentation for alternative uses
  import fnmatch, glob, traceback, errno, sys, atexit, locale, imp, stat

Traceback (most recent call last):

  File "/home/stdiff/.local/bin/imagine", line 8, in <module>
    sys.exit(imagine_cmd())

  File "/usr/lib/python3/dist-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)

  File "/usr/lib/python3/dist-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)

  File "/usr/lib/python3/dist-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)

  File "/usr/lib/python3/dist-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)

  File "/usr/lib/python3/dist-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)

  File "/home/stdiff/.local/lib/python3.10/site-packages/imaginairy/cmds.py", line 223, in imagine_cmd
    imagine_image_files(

  File "/home/stdiff/.local/lib/python3.10/site-packages/imaginairy/api.py", line 114, in imagine_image_files
    for result in imagine(

  File "/home/stdiff/.local/lib/python3.10/site-packages/imaginairy/api.py", line 313, in imagine
    is_nsfw_img = is_nsfw(img, x_sample)

  File "/home/stdiff/.local/lib/python3.10/site-packages/imaginairy/safety.py", line 23, in is_nsfw
    _, has_nsfw_concept = safety_checker(

  File "/home/stdiff/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)

  File "/home/stdiff/.local/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)

  File "/home/stdiff/.local/lib/python3.10/site-packages/diffusers/pipelines/stable_diffusion/safety_checker.py", line 39, in forward

    special_cos_dist = cosine_distance(image_embeds, self.special_care_embeds).cpu().numpy()

TypeError: Got unsupported ScalarType BFloat16
```

**Comment by brycedrennan** [comment]
okay 1.6.2 deployed with fix @iscopan

**Comment by brycedrennan** [comment]
Okay I'll give it a more thorough look over tonight.

**Comment by brycedrennan** [comment]
@iscopan I just released 2.0.1, which I think contains a fix for this last error you mentioned. 🤞

**Comment by brycedrennan** [comment]
As i think this is fixed, closing until I hear otherwise.
